### ISSUE TITLE
RFC: Vendor Bundling

### ISSUE THREAD

**Issue by delambo** [NYTimes/kyt#315]
## Summary

Since vendor assets change less frequently than application code it is typical to extract and bundle them separately in webapps. Out of the box, kyt should have an opinionated strategy for bundling a user's vendor assets.

## Motivation

Internally, and in the [redux starter kyt poc](https://github.com/NYTimes/kyt-starter-universal/compare/master...julianvmodesto:master#diff-082dd42d090dcb803fca0e5cd6428debR35) we separately bundled vendor libraries. There is a very nice writeup on the technique that we used by @tizmagik [here](https://jeremygayed.com/dynamic-vendor-bundling-in-webpack-528993e48aab#.4kh1t8qam). Outside of these use cases, vendor bundling is an established pattern for advanced webapps. Since kyt's goal is to make it easy to setup and write advanced webapps, vendor bundling should be a core opinion that is baked into the framework.

## Detailed design

Vendor bundling is only relevant in a client build so this change should only need to be implemented in the webpack client development and production environment configurations. Since one configuration should work across both environments a utility should be used to export the common functionality. The recently updated implementation outlined by @tizmagik in his [post](https://jeremygayed.com/dynamic-vendor-bundling-in-webpack-528993e48aab#9d1b) is the proposed implementation. Specifically, the `CommonsChunkPlugin` and `WebpackMd5Hash` plugins should be included in the `plugins` section of the kyt client webpack configurations.

In addition to the plugins, kyt could export an ordered bundle manifest during the `dev` and `build` commands. For more, refer to Unresolved Questions (1).

## Drawbacks

1. It's hard to override this feature. Since the implementation mostly adds webpack plugins, it would be burndensome for a user to find and remove the chunking plugins.

2. Depending on how we work out Unresolved Questions (1), this feature could add a lot of complexity in userland when it is time to sort and load bundles.

## Alternatives

1. Let users manage their own bundling in the kyt.config.js `modifyWebpackConfig`.

## Unresolved questions

1. Could kyt export an ordered bundle asset manifest? Currently, kyt uses the AssetsWebpackPlugin to export an unordered bundle manifest. Users are expected to import the manifest into their projects and manually load the named bundles. Since the vendor bundle needs to be loaded before the entry bundles and kyt knows the vendor bundle name, kyt could export the bundle names in the order that they need to load. A potential solution would be to create an internal webpack plugin that works similarly to the AssetsWebpackPlugin but uses knowledge of kyt conventions to order the bundles.

2. Will this work with the html-webpack-plugin/kyt-starter-static?

**Comment by tizmagik** [comment]
+1! I'd love to see this solved in kyt core. I think one way we could get around some of the drawbacks you outlined are to wrap up all the bundles necessary into a kyt-specific bundle. Imagine a Webpack config that looks like this:

```js
const webpack = require('webpack');
const KytBundlerPlugin = require('webpack-kyt-bundler-plugin');

const webpackConfig = {
  entry: { main: './app.js' },
  output: { 
    filename: '[name]-[chunkhash].js', // note *chunkhash* used here
    chunkFilename: '[name]-[chunkhash].js', 
  },
  plugins: [
    new KytBundlerPlugin()
  ]
}
```

Which could include:
  - Vendor+Manifest chunking
  - MD5 Hash names
  - Assets plugin
  - HTML plugin

All abstracted away from the user internally to kyt. We may want to split these up into multiple kyt-flavored plugins, and we may or may not even publish them independently of `kyt`, but the point is it could achieve a few things:

1. Abstract away the vendor/manifest chunking from the user
1. Ensure proper vendor chunking
1. Ensure proper chunk load ordering
1. Make it easy for users to override this behavior by stripping out this one (or two) kyt-flavored plugins.

---

BTW, I think you have a typo in your post:
>Since vendor assets change **~~more~~ less** frequently than application code

**Comment by delambo** [comment]
I just verified that Unresolved Question (2) is not a problem. I tested the static starter-kyt and the html-webpack-plugin did load the bundles in the correct order.

**Comment by jaredmcdonald** [comment]
regarding Drawbacks (1), could we provide an option in `kyt.config.js` to disable it? (the opinionated default would be to include it)

**Comment by delambo** [comment]
I'm going to start POC'ing this. Since webpack asset hashing is broken, we created a script for our internal projects that correctly hashes bundles. I think the DLL plugin might be able to replace our script so I'm going to check that out first and hopefully it will have the side benefit of speeding up our build times as reported in #216.

**Comment by delambo** [comment]
Vendor bundling was implemented and published in version 0.6.0. Check out the [migration guide](https://github.com/NYTimes/kyt/blob/master/docs/migration-guides/0.5-0.6.md).
